### ISSUE TITLE
NodeID is 0

### ISSUE THREAD

**Issue by ziflex** [mafredri/cdp#64]
Hi!

First of all, I want to thank you for your incredible work of bringing CDP to go land. That's fantastic.

Second, I'm working on scraping tool where I'm trying to use the package as an access to JS-based web pages. During the use, I've noticed, sometimes ```DescribeNode``` (but, ```BackendId``` is not empty) return a node with empty ```NodeId``` field and as a result I have to keep the id returned from ```QuerySelector```/```QuerySelectorAll``` methods.

I'm wondering whether this behavior is by design?

**Comment by mafredri** [comment]
Hi @ziflex,

I'm not sure I'm the right person to answer this question, my understanding of Nodes and BackendNodes is a bit foggy.

I can answer one thing though:

> To answer your question if it's by design

It would seem so. But to understand why, I think it's best if I provide some links to relevant resources than try to explain it (broken telephone and all that):

[How are NodeIDs expected to be used?](https://groups.google.com/a/chromium.org/forum/#!topic/headless-dev/A4kNwvX77Bs/discussion) (headless-dev)

Alex Clarke wrote:
> From what I understand most of those commands where implemented for the DevTools debugger and are not considered useful for scripting with.  Generally I wouldn't recommend using NodeID in scripts because various (undocumented) things can invalidate them and the implementation of which could change.
>
> BackendNode IDs are much more stable although not every command supports them yet.

[Retrieve node from nodeId](https://github.com/ChromeDevTools/devtools-protocol/issues/20)

pavelfeldman wrote:
> Present DOM domain serves the 'inspection' goal, not scripting or automation. If you have nodeId in hand, it means that all the node information has already been sent to you in one of the responses or notifications.

[Add BackendNodeId param to the DOM.querySelectorAll method](https://github.com/ChromeDevTools/devtools-protocol/issues/50)

pavelfeldman wrote:
> [...] I would still encourage you to traverse frames and run querySelectorAll in them. The reason we can't do querySelectorAll by backendNodeId is that the nodes we return would need to have nodeIds, which means full blown DOM domain functionality we would want to work around.

I hope this can be of help!

**Comment by ziflex** [comment]
Thanks a lot! Maybe it makes sense to add this information to documentation?

**Comment by mafredri** [comment]
Do you have a suggestion where it could be added and makes sense? Primarily this library is about exposing the protocol for Go programming but I also think it would be worthwhile to decrease the barrier of entry as the protocols own documentation is quite lacking. But I also don’t want to distract from this library’s own documentation.

Maybe it would make sense to link to a FAQ in the readme.

**Comment by ziflex** [comment]
@mafredri  yeah, I think FAQ or some notes regarding imperfectness of the protocol :) 
I'm currently building [web scraping system](https://github.com/MontFerret/ferret) based on your library, so I'm very open to help you make it better :)

**Comment by ziflex** [comment]
Btw, I wonder, if the NodeId is not reliable, why is it used so widely? Wouldn't be easier just do eval for that?

**Comment by mafredri** [comment]
How do you mean widely used? NodeID was born pretty early on I believe and was optimized for the DevTools (UI) use-case, e.g. DevTools listens to all dom events and keeps an internal representation that is constantly updating and when a response contains a NodeID DevTools knows it already has the relevant data.

But yeah, eval is probably better for scripting since you're often working with "new" data (e.g. has no NodeID yet).

**Comment by ziflex** [comment]
By widely use I mean that they are in almost all API calls even though they are stated as unstable. That's confusing, but I guess it's a historical reason.
I'm still exploring the API, and trying to figure out what is the best way of dealing with it.
As far as I know, there is no something like special Node struct which would track it's own state, like a special Node binding. (For example, you got a Node from DescribeNode or something, and every time its content or one of the attributes changes triggered by the page events, that would be reflected in this  binding).
Even though, the idea of CDP (the protocol itself) is great, it feels that there are many gaps in terms of usability.
